### PR TITLE
transaction-pool: drop unpropagable txs if local node cant author blocks

### LINKED ISSUE #8045: Only block authors should store non-propagateable transactions in the pool.
Currently if the node is not a block author, and the `TransactionValidity` ends up with `propagate=false` there is nothing we can do with such transaction. It won't be propagate to other peers and since we are not authoring blocks, it will never end up in any block.
It simply sits in the pool occupying a slot and waiting to run out of it's `longevity`.

Instead the pool should never accept such transactions if we know the node is not authoring blocks.

### PR DESCRIPTION
**Pull request by andresilva**
Fixes #8045.

polkadot companion: https://github.com/paritytech/polkadot/pull/2385

### CODE REVIEW

File: client/transaction-pool/graph/src/validated_pool.rs [hunk 1]
```diff
@@ -90,9 +90,25 @@ pub type ValidatedTransactionFor<B> = ValidatedTransaction<
 	<B as ChainApi>::Error,
 >;
 
+/// A closure that returns true if the local node can author blocks.
+pub struct CanAuthor(Box<dyn Fn() -> bool + Send + Sync + 'static>);
```
Comment (tomusdrw) [inline client/transaction-pool/graph/src/validated_pool.rs:94]:
```suggestion
pub struct CanAuthor(Box<dyn Fn() -> bool + Send + Sync>);
```
`'static'` is implied from the `Box`, no?

File: client/transaction-pool/graph/src/validated_pool.rs [hunk 2]
```diff
@@ -90,9 +90,25 @@ pub type ValidatedTransactionFor<B> = ValidatedTransaction<
...
+
+impl From<bool> for CanAuthor {
+	fn from(can_author: bool) -> Self {
+		CanAuthor(Box::new(move || can_author))
+	}
+}
+
+impl From<Box<dyn Fn() -> bool + Send + Sync + 'static>> for CanAuthor {
+	fn from(can_author: Box<dyn Fn() -> bool + Send + Sync + 'static>) -> Self {
```
Comment (tomusdrw) [inline client/transaction-pool/graph/src/validated_pool.rs:103]:
```suggestion
impl From<Box<dyn Fn() -> bool + Send + Sync>> for CanAuthor {
	fn from(can_author: Box<dyn Fn() -> bool + Send + Sync>) -> Self {
```

File: primitives/transaction-pool/src/error.rs [hunk 3]
```diff
@@ -60,6 +60,9 @@ pub enum Error {
 	#[error("Transaction couldn't enter the pool because of the limit")]
 	ImmediatelyDropped,
 
+	#[error("Transaction cannot be propagated and local node does not author blocks")]
+	Unactionable,
```
Comment (tomusdrw) [inline primitives/transaction-pool/src/error.rs:64]: Might be good to add RPC translation for that error too

**Review by tomusdrw — APPROVED** [review]
lgtm!

**Comment by andresilva** [comment]
This is still missing tests but I'll add one in a bit.

**Review by bkchr — APPROVED** [review]

**Pull request merged**